bpfassets/perf_event/perf_event.c

```diff
@@ -206,30 +207,26 @@ int kprobe__finish_task_switch(struct pt_regs *ctx, struct task_struct *prev)
 
     // store process metrics
     struct process_metrics_t *process_metrics;
+    process_metrics = processes.lookup(&prev_pid);
+    if (process_metrics != 0)
+    {
+        // update process time
+        process_metrics->process_run_time += on_cpu_time_delta;
+
+        process_metrics->cpu_cycles += on_cpu_cycles_delta;
+        process_metrics->cpu_instr += on_cpu_instr_delta;
+        process_metrics->cache_miss += on_cpu_cache_miss_delta;
+    }
+
     process_metrics = processes.lookup(&cur_pid);
     if (process_metrics == 0)
     {
         process_metrics_t new_process = {};
         new_process.pid = cur_pid;
         new_process.cgroup_id = cgroup_id;
-        new_process.process_run_time = on_cpu_time_delta;
         bpf_get_current_comm(&new_process.comm, sizeof(new_process.comm));
-
-        new_process.cpu_cycles = on_cpu_cycles_delta;
-        new_process.cpu_instr = on_cpu_instr_delta;
-        new_process.cache_miss = on_cpu_cache_miss_delta;
-
         processes.update(&cur_pid, &new_process);
```

Comment:
This also should be the `prev_pid` and we should add the initial values
```
        new_process.process_run_time = on_cpu_time_delta;
        new_process.cpu_cycles = on_cpu_cycles_delta;
        new_process.cpu_instr = on_cpu_instr_delta;
        new_process.cache_miss = on_cpu_cache_miss_delta;
        processes.update(&prev_pid, &new_process);
```

Reply:
If we do not add the initial values, we will loose it

Reply:
The full example of the code:
```c
int kprobe__finish_task_switch(struct pt_regs *ctx, struct task_struct *prev)
{
    // u32 cur_pid = bpf_get_current_pid_tgid() >> 32;
    struct task_struct *task = (struct task_struct *)bpf_get_current_task();
    u32 cur_pid = task->pid;    
#ifdef SET_GROUP_ID
    u64 cgroup_id = bpf_get_current_cgroup_id();
#else
    u64 cgroup_id = 0;
#endif
    u32 prev_pid = prev->pid;

    u64 cur_ts = bpf_ktime_get_ns();
    u32 cpu_id = bpf_get_smp_processor_id();
    u64 on_cpu_time_delta = get_on_cpu_time(cur_pid, prev_pid, cur_ts);
    u64 on_cpu_cycles_delta = get_on_cpu_cycles(&cpu_id);
    u64 on_cpu_ref_cycles_delta = get_on_cpu_ref_cycles(&cpu_id);
    u64 on_cpu_instr_delta = get_on_cpu_instr(&cpu_id);    
    u64 on_cpu_cache_miss_delta = get_on_cpu_cache_miss(&cpu_id);
    u64 on_cpu_avg_freq = get_on_cpu_avg_freq(&cpu_id, on_cpu_cycles_delta, on_cpu_ref_cycles_delta);

    // store process metrics
    struct process_metrics_t *process_metrics;
    process_metrics = processes.lookup(&prev_pid);
    if (process_metrics == 0)
    {
        process_metrics_t new_process = {};
        new_process.pid = prev_pid;
        new_process.cgroup_id = cgroup_id;
        bpf_probe_read_kernel(&new_process.comm, sizeof(new_process.comm), &prev->comm);

        new_process.process_run_time = on_cpu_time_delta;
        new_process.cpu_cycles = on_cpu_cycles_delta;
        new_process.cpu_instr = on_cpu_instr_delta;
        new_process.cache_miss = on_cpu_cache_miss_delta;

        processes.update(&prev_pid, &new_process);
    }
    else
    {
        // update process time
        process_metrics->process_run_time += on_cpu_time_delta;
        process_metrics->cpu_cycles += on_cpu_cycles_delta;
        process_metrics->cpu_instr += on_cpu_instr_delta;
        process_metrics->cache_miss += on_cpu_cache_miss_delta;
    }

    return 0;
}
```

Reply:
```c
struct process_metrics_t *process_metrics;
    process_metrics = processes.lookup(&prev_pid);
    if (process_metrics != 0)
    {
        // update process time
        process_metrics->process_run_time += on_cpu_time_delta;

        process_metrics->cpu_cycles += on_cpu_cycles_delta;
        process_metrics->cpu_instr += on_cpu_instr_delta;
        process_metrics->cache_miss += on_cpu_cache_miss_delta;
    }

    process_metrics = processes.lookup(&cur_pid);
    if (process_metrics == 0)
    {
        process_metrics_t new_process = {};
        new_process.pid = cur_pid;
        new_process.cgroup_id = cgroup_id;
        bpf_get_current_comm(&new_process.comm, sizeof(new_process.comm));
        processes.update(&cur_pid, &new_process);
    }
```

Reply:
Ok! I believe your current approach is more effective as it correctly set the cgroup id.